pkg/cluster/node.go

```diff
@@ -4,71 +4,78 @@
 package cluster
 
 import (
+	"context"
 	"fmt"
 	"io"
 	"log"
 	"net"
+	"net/http"
 	"time"
 
+	"github.com/google/uuid"
 	"github.com/hashicorp/serf/serf"
 	"github.com/mitchellh/hashstructure/v2"
 	"github.com/nsqio/go-nsq"
 	"go.uber.org/zap"
 )
 
+const (
+	eventHandlerWorkers = 1 // TODO: should this be 1?
+	queueSize           = 100
+)
+
 // Node is configured per host in the cluster.
 type Node struct {
 	logger *zap.SugaredLogger
 
 	// serf settings
 	serfServer *serf.Serf
 	events     chan serf.Event
-	nodefinder NodeFinder
 	upCh       chan bool
 
 	// nsq settings
 	bus            *bus
 	busChannelName string
 	producer       *nsq.Producer
+	id             string
+	httpClient     *http.Client
+	maxRetries     int
 }
 
-func NewNode(config Config, nodeFinder NodeFinder, logger *zap.SugaredLogger) (*Node, error) {
+func NewNode(ctx context.Context,
+	config Config,
+	getAddr func(ctx context.Context, nodeID string) (string, error),
+	getNodes func(context.Context) ([]string, error),
+	timeout time.Duration,
+	logger *zap.SugaredLogger,
+	httpClient *http.Client,
+	maxRetries int,
+) (*Node, func(), error) {
 	var err error
 
-	if logger == nil {
-		logger = zap.NewNop().Sugar()
-	}
-
-	if nodeFinder == nil {
-		panic(fmt.Errorf("nodefinder not set"))
-	}
-
 	node := &Node{
 		logger:     logger,
-		nodefinder: nodeFinder,
 		upCh:       make(chan bool),
+		id:         uuid.NewString(),
+		httpClient: httpClient,
 	}
 
-	node.bus, err = newBus(config, logger)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create nsq bus: %w", err)
-	}
-	go func() {
-		e := node.bus.start()
-		if e != nil {
-			panic("can not start nsq bus")
-		}
-	}()
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
I believe cancelling contexts is always necessary. This function doesn't cancel if everything returns successfully.